File: lib/ftw_realtime/marketplace/user_setting.ex
defmodule FtwRealtime.Marketplace.UserSetting do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "user_settings" do
    field :notifications_email, :boolean, default: true
    field :notifications_push, :boolean, default: true
    field :notifications_sms, :boolean, default: false
    field :appearance_theme, :string, default: "light"
    field :language, :string, default: "en"
    field :timezone, :string, default: "America/Chicago"
    field :privacy_profile_visible, :boolean, default: true
    field :privacy_show_rating, :boolean, default: true

    belongs_to :user, FtwRealtime.Marketplace.User

    timestamps(type: :utc_datetime)
  end

  def changeset(setting, attrs) do
    setting
    |> cast(attrs, [
      :user_id,
      :notifications_email,
      :notifications_push,
      :notifications_sms,
      :appearance_theme,
      :language,
      :timezone,
      :privacy_profile_visible,
      :privacy_show_rating
    ])
    |> validate_required([:user_id])
    |> validate_inclusion(:appearance_theme, ~w(light dark system))
    |> unique_constraint(:user_id)
    |> foreign_key_constraint(:user_id)
  end
end

File: src/main/kotlin/com/strata/ftw/domain/entity/UserSetting.kt
package com.strata.ftw.domain.entity

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "user_settings", uniqueConstraints = [UniqueConstraint(columnNames = ["user_id"])])
class UserSetting(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid")
    var id: UUID? = null,

    @Column(name = "notifications_email")
    var notificationsEmail: Boolean = true,

    @Column(name = "notifications_push")
    var notificationsPush: Boolean = true,

    @Column(name = "notifications_sms")
    var notificationsSms: Boolean = false,

    @Column(name = "appearance_theme")
    var appearanceTheme: String = "light",

    var language: String = "en",

    var timezone: String = "America/Chicago",

    @Column(name = "privacy_profile_visible")
    var privacyProfileVisible: Boolean = true,

    @Column(name = "privacy_show_rating")
    var privacyShowRating: Boolean = true,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    var user: User? = null,

    @Column(name = "user_id", insertable = false, updatable = false)
    var userId: UUID? = null,

    @CreationTimestamp
    @Column(name = "inserted_at", updatable = false)
    var insertedAt: Instant? = null,

    @UpdateTimestamp
    @Column(name = "updated_at")
    var updatedAt: Instant? = null
)
